Handle missing bsp info better

Arthurm1 opened the pull request:
Gradle BSP doesn't support Scala and also doesn't supply a DisplayName.  BaseDirectory is also optional in the spec.

Not sure if this is the best way to dynamically create a DisplayName.  Currently the ScalaTarget/JavaTarget data is extracted as required.  It would probably be better to create classes to hold the info and then create unique DisplayNames for each target as it's not guaranteed that the BSP server supplied DisplayNames will be unique.  But possibly not worth it though as I'd hope Gradle would support a DisplayName in next versions.

This is to get the Doctor and Build Info working and tree view partially working and to stop logging of stack traces when BSP server doesn't support certain requests

tgodzik reviewed (changes requested):
Looks like the Bill server that is used for some of the tests doesn't have something properly set, it's probably one of the capabilities.

Could you take a look?

Arthurm1 commented on metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala:
```diff
@@ -105,6 +111,22 @@ object MetalsEnrichments
+          else
+            buildTarget.getId().getUri()
+
+        name.replaceAll("[^a-zA-Z0-9]+", "-")
```
@tgodzik Should I change this so the non A->Z chars are _always_ replaced with `-`, not just when `DisplayName` is not supplied?
The issue is that in `FileDecoderProvider` Metals uses the display name to build up a URI to then display build target info to the user.  If there is a non URI valid character then this URI is built incorrectly.  You can see that issue also in https://github.com/scalameta/metals/pull/5816.
The other option would be to change the way build target info is displayed but that would probably involve changing each individual clients (vscode, vim etc) as well.

Arthurm1 commented:
@tgodzik I've changed the test server setup to always supply `dataKind` as `scala` if `data` is `ScalaBuildTarget` in `BuildTarget`.

The spec is [here](https://build-server-protocol.github.io/docs/specification#:~:text=dataKind%3F%3A%20BuildTargetDataKind) and says both `dataKind` and `data` are optional but it makes sense that if a server populates one then it has to populate the other.
In the Gradle BSP case the server was supplying `jvm` and Metals was decoding it as though it were `scala`.
In the test server cases, `dataKind` was `null` which I've changed to `scala`.

Arthurm1 commented:
WorksheetLspSuite fails but that looks as though it's down to artifact downloading?  It just hangs when I try and run it locally on Windows so I can't check

tgodzik reviewed (approved):
LGTM!

tgodzik commented:
I rebased the PR, will merge tomorrow.
